### PR TITLE
Show/hide toolbar and handles based on device kind

### LINKED ISSUE #31629: The logic for showing/hiding text selection handles should live in platform-specific code
Different platforms have different criteria for showing/hiding the toolbar and selection handles in text fields. The current direction is to make the platform-independent widgets (`RenderEditable` and `EditableText`) expose the necessary APIs but they shouldn't make the decision on when to show or hide something. The platform-specific widgets (`material.TextField` and `cupertino.TextField`) should make those decisions and call the APIs provided by platform-independent widgets.

## BREAKING CHANGE

One of the bits that today is not inline with our direction is the showing/hiding of text selection handles. Currently that's done in `EditableText` and we should move it to `material/cupertino`'s `TextField`.

### PR DESCRIPTION
**Pull request by mdebbar**
## Description

When the user is interacting with a text field through a mouse, we don't want to show the toolbar nor the selection handles. They should only be shown when the user is interacting through a touch screen.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [ ] No, this is *not* a breaking change.

Fixes https://github.com/flutter/flutter/issues/31629

## TODO
- [x] Implement for material text field.
  - [x] Write tests.
- [x] Implement for cupertino text field.
  - [x] Write tests.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/editable_text.dart]: This is just e a editabletext. It shouldn't know so much about whether to show the toolbar based on the platform and pointer type etc etc. Just let its parent call or not call showToolbar() based on the appropriate platform+device information.

**Comment by xster** [comment]
I don't have a good idea of how useful knowing the kind will be in general for all gestures. cc @dkwingsmt, @gspencergoog. But if it's all very text fields / TextSelectionGestureDetector specific, you can just add another Listener to TextSelectionGestureDetector that specifically check for pointer kinds coming in and out and pass that back out via a separate callback. For instance, you can pass a pointers tracker into TextSelectionGestureDetector while constructing, and on callbacks of the other callbacks, its parent can ask the tracker it passed in if it's all exclusively tap or mouse and do something special if it's a mix (though it'd be weird that you get a callback at all in that case).

File: packages/flutter/lib/src/widgets/editable_text.dart
Comment (mdebbar) [reply packages/flutter/lib/src/widgets/editable_text.dart]: Agreed. It seems that the logic is still spread across editable text and text field now. I did a clean up and put all the show/hide logic in the text field. PTAL.

**Comment by mdebbar** [comment]
@xster I like the idea of a having a tracker for device kind. I just updated the PR with the changes you recommended. Also added a bunch of tests to make sure the behavior toolbar/handles doesn't break.

File: packages/flutter/lib/src/widgets/gesture_detector.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/gesture_detector.dart]: I think it's a slightly 'misleading' usage of ValueNotifier since there are no usages of the Listenable interface or listeners etc. Do you intend to expand its usage in the future?

Also, since we were saying we're going this route to avoid having to design a whole generic API around tracking the kind type (since we think it's only useful in the text selection domain), these knowledge shouldn't leak outside of the text selection classes. 

I would just create a listener that composes a Listener above the TextSelectionGestureDetector that doesn't participate in the arena or gestures etc and that only keeps track of the last primary pointer's kind and reports back to the text fields that injected the tracker to the TextSelectionGestureDetector.

Comment (mdebbar) [reply packages/flutter/lib/src/widgets/gesture_detector.dart]: Ahh, I misunderstood your previous comment :P
I switched now to using `Listener` and it's much cleaner indeed. I'll push it right away.

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: You should decide on this, though while we want to really keep this simple, this may be a bit too simple :D. I think you might still want a little bit of the pointer tracking here. Unless you think the pointer events will never be in a situation where the kinds will mix.

Comment (mdebbar) [reply packages/flutter/lib/src/cupertino/text_field.dart]: This is by design. We only want to take into account the actions that "trigger" the text field. All actions that "trigger" the text field start with a PointerDown event (tap, long press, drag, long-press-drag, etc).

See the [internal Q/A](https://yaqs.googleplex.com/eng/q/4519774183489536) for the opinion of the Material team on this.

Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: This is really an edge case, but I was asking what is supposed to happen if you held down and dragged a finger on Android, selecting some words, and right before you lift the finger, there was also a mouse click somewhere. 

By your rule above, the 'trigger' is the touch, therefore it should show the toolbar, but by the current implementation, the last kind is mouse and there won't be a toolbar.

File: packages/flutter/test/cupertino/text_field_test.dart
Comment (xster) [inline packages/flutter/test/cupertino/text_field_test.dart]: We should also check that long press always produces collapsed selection (thus no handles)

**Review by xster — COMMENTED** [review]
This is much better now. Thanks!

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (mdebbar) [reply packages/flutter/lib/src/cupertino/text_field.dart]: There are cases where things bubble in then out, but many selection changes originate from render editable (e.g. using a keyboard on Android, or semantics).

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (mdebbar) [reply packages/flutter/lib/src/cupertino/text_field.dart]: You're right. A pen (stylus) should also trigger the selection overlay because it's still interacts through the touch screen. Do you think `invertedStylus` should also trigger the selection overlay? I don't have a pen at the moment to try this.

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (mdebbar) [reply packages/flutter/lib/src/cupertino/text_field.dart]: The reason I formatted them this way is because the first two return false while the 3rd and 4th return true. I'll add an empty line between all of them.

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: This might need a comment above. Isn't long press selections on iOS always collapsed?

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (xster) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Is this an API breaking change? i.e. if I previously used EditableText as is, it won't show handles anymore after I flutter upgrade?

File: packages/flutter/lib/src/widgets/text_selection.dart
Comment (xster) [inline packages/flutter/lib/src/widgets/text_selection.dart]: Add a bit more doc here and with `hide` to describe their differences

**Review by xster — APPROVED** [review]
LGTM

**Comment by dnfield** [comment]
@mdebbar will you be addressing this feedback?

File: packages/flutter/lib/src/cupertino/text_field.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/cupertino/text_field.dart]: It seems like to @dkwingsmt and me that a better way to do this would be to add the pointer kind to the details for the gesture detector callbacks, and look at the device kind when needed.  Then you wouldn't need the extra listener, or to maintain a cache of the device kind.

Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: I gave him the opposite advice previously :D

https://github.com/flutter/flutter/pull/29683#issuecomment-475003862

because I wasn't sure where you guys wanted to go with altering signatures everywhere. If you guys were going there anyway, then what you're suggesting makes sense.

Comment (gspencergoog) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Oh, sorry, I didn't notice that.  Yeah, we're going to have to go there anyhow, since there are going to be other cases of needing to distinguish.  The good thing is that it won't alter a lot of the signatures if there are details objects already, since we can just add it to the details that are already provided.

Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Sounds reasonable. Sorry for the back and forth @mdebbar.

Comment (mdebbar) [reply packages/flutter/lib/src/cupertino/text_field.dart]: @gspencergoog thanks for chiming in. @xster it's all good :)

I'll have the necessary changes ready for review by EOD.

**Comment by mdebbar** [comment]
@xster @gspencergoog I updated the PR. Some tests are still failing but I would like to get a review while I fix those tests. From a quick glance, they seem to be newly added tests for `EditableText` showing the handles. Since the logic for showing/hiding the handles is now in `TextField`, I'll port those tests over.

File: packages/flutter/lib/src/gestures/recognizer.dart
Comment (mdebbar) [inline packages/flutter/lib/src/gestures/recognizer.dart]: This isn't ideal. If you have better ideas on how to do it, I'm all ears :)

Comment (xster) [reply packages/flutter/lib/src/gestures/recognizer.dart]: Ah, this is unfortunate. Though this event isn't entirely correct either right? Since you're sending something from the past and the past event also has a timeStamp. 

I haven't validated it but I suspect each onLongPressDown is preceded by a onTapDown or something. From the TextSelectionGestureDetector's perspective, it may be enough?

**Review by xster — COMMENTED** [review]
Thanks for changing this around. Sorry again for the back and forth.

**Comment by goderbauer** [comment]
@mdebbar Are you still planning to get back to this? The PR is getting kinda old...

File: packages/flutter/lib/src/gestures/recognizer.dart
Comment (mdebbar) [reply packages/flutter/lib/src/gestures/recognizer.dart]: Unfortunately, this is how the tap recognizer works. It doesn't send the tap down event until either a tap up occurs, or the deadline is exceeded. And in that case, the tap down is always from the past (even before my change). All my change does is it adds the `kind` field to the event.

**Pull request merged**
